examples/02_intermediate/wikimedia.ipynb

```diff
@@ -0,0 +1,513 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "4104ae2f-defd-41a3-85ae-30f2c2271b3e",
+   "metadata": {},
+   "source": [
+    "Now, let's explore writing a _historical_ adapter to read the same data stream. The Wikimedia API offers the following query for historial data:\n",
+    "\n",
+    "```\n",
+    "URL = f'https://stream.wikimedia.org/v2/stream/recentchange?since={start_date}'\n",
+    "```\n",
+    "\n",
+    "where `{start_date}` is a string representing the start of the period we want to read from. \n",
+    "\n",
+    "The stream history for Wikimedia data is not kept indefinitely. Depending on the particular stream configuration, there will likely be between 7 and 31 days of history available. We will choose a past date to start reading data, and stop after 23h of events."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "7bc632d9-b47c-4f3d-87fc-3364f90dab4c",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from datetime import datetime, timezone\n",
+    "\n",
+    "start_time = datetime(2024, 5, 8, 0, 0, 0)\n",
+    "end_time = datetime(2024, 5, 8, 23, 0, 0, tzinfo=timezone.utc)"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "id": "80702f29-a9c7-45e8-8e77-7cc717f4858a",
+   "metadata": {},
+   "source": [
+    "To read this data, we will use a [Pull input adapter](https://github.com/Point72/csp/wiki/Write-Historical-Input-Adapters#pullinputadapter). However, since our historical data can be downloaded into a csv file, we will use the existing implementation of a [`CSVReader` adapter](https://github.com/Point72/csp/blob/main/csp/adapters/csv.py) from CSP."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "b513164c-9e2f-44c9-a7c2-9a93357f4cb9",
+   "metadata": {},
+   "source": [
+    "Again, we will use a Perspective widget to display the results of our historical data. This time, the widget will immediately show the historical data."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "555b24f7-f765-4711-90a9-46db5555a50f",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import json\n",
+    "from sseclient import SSEClient as EventSource, Event\n",
+    "from perspective import PerspectiveWidget, Table, Plugin\n",
+    "import pandas as pd\n",
+    "\n",
+    "start_date = datetime.isoformat(start_time)\n",
+    "end_date = datetime.isoformat(end_time)\n",
+    "\n",
+    "URL = f'https://stream.wikimedia.org/v2/stream/recentchange?since={start_date}'\n",
+    "\n",
+    "# Initial data\n",
+    "data = pd.DataFrame({\n",
+    "    \"servername\": [\"\"],\n",
+    "    \"time\": [datetime.utcnow()],\n",
+    "    \"event\": [\"\"],\n",
+    "    \"servername_count\": [0],\n",
+    "})\n",
+    "\n",
+    "raw_data = pd.DataFrame(dict([]))\n",
+    "table = Table(data, index=\"servername\")\n",
+    "servernames = dict([])\n",
+    "for event in EventSource(URL):\n",
```

Comment:
since wikimedia has an endpoint to get historical data, this example should actually have a PullAdapter written which gets data from thsi url.
I wouldn't go from this -> CSV and then use CSVReader.  
Then the example can use both realtime and historical adapter interchangeably and wired into the perspective widget and the filter nodes

Reply:
I can write another adapter here, my choice for the CSV adapter was to highlight that it exists, since we don't really document it elsewhere in the wiki. Would you prefer that we do just the custom adapter, or both the custom and the csv one?

Reply:
I think this is a good opportunity to demonstrate writing both a push and pull adapter and have them serve up the same exact type of data, which makes them interchangeable